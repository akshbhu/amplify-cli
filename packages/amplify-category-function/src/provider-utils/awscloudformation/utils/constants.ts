export const categoryName = 'function';
export const serviceName = 'Lambda';
export const functionParametersFileName = 'function-parameters.json';
export const provider = 'awscloudformation';