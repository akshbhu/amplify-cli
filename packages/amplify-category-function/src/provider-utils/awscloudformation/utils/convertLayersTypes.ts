import { ExternalLayer, LambdaLayer, ProjectLayer } from 'amplify-function-plugin-interface';
import { convertProjectLayer } from './layerArnConverter';

const LAYER_ARN_KEY = 'Fn::Sub';

// convert project layers to external when changing env
// only latest version layer will remain a project layer
export const convertProjectLayersToExternalLayers = (lambdaLayers: LambdaLayer[], envName: string): LambdaLayer[] => {
  const modifiedLambdaLayers: LambdaLayer[] = [];
  lambdaLayers.forEach(layer => {
    if (layer.type === 'ProjectLayer' && layer.env !== envName && !layer.isLatestVersionSelected) {
      const convertLayer: ExternalLayer = {
        type: 'ExternalLayer',
        arn: convertProjectLayer(layer, envName),
      };
      modifiedLambdaLayers.push(convertLayer);
    } else {
      const convertLayer: ProjectLayer = {
        type: 'ProjectLayer',
        resourceName: (layer as ProjectLayer).resourceName,
        version: (layer as ProjectLayer).version,
        isLatestVersionSelected: true,
        env: envName,
      };
      modifiedLambdaLayers.push(convertLayer);
    }
  });
  return modifiedLambdaLayers;
};

//This functions convert external layers to project layers if they belong to the env
// exmaple Arn to convert external layers to project Layers when changing env
// "Fn::Sub": "arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:layer:buildlayers8mytestinglayer1-dev:2"
export const covertExternalLayersToProjectLayers = (lambdaLayers: LambdaLayer[], envName: string): LambdaLayer[] => {
  const modifiedLambdaLayers: LambdaLayer[] = [];
  lambdaLayers.forEach(layer => {
    if (layer.type === 'ExternalLayer' && layer.arn.hasOwnProperty(LAYER_ARN_KEY)) {
      const layerArn = layer.arn[LAYER_ARN_KEY];
      const layerNameWithEnv = layerArn.split(':')[layerArn.split(':').length - 2];
      const layerVersion = parseInt(layerArn.split(':')[layerArn.split(':').length - 1], 10);
      const layerName = layerNameWithEnv.split('-')[0];
      const layerEnv = layerNameWithEnv.split('-')[1];
      if (envName !== layerEnv) {
        modifiedLambdaLayers.push(layer);
      } else {
        modifiedLambdaLayers.push({
          type: 'ProjectLayer',
          resourceName: layerName,
          version: layerVersion,
          isLatestVersionSelected: false,
          env: layerEnv,
        });
      }
    }
  });
  return modifiedLambdaLayers;
};
