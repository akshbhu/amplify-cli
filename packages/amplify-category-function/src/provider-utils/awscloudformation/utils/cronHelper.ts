import inquirer from 'inquirer';
inquirer.registerPrompt('datetime', require('inquirer-datepicker'));
import { CronBuilder } from '../utils/cronBuilder';

var Days = {
    MON: '1',
    TUE: '2',
    WED: '3',
    THU: '4',
    FRI: '5',
    SAT: '6',
    SUN: '7',
  };

export async function minuteHelper(context : any){
    const minuteQuestion = {
        type: 'input',
        name: 'minutes',
        message: 'Enter the rate in mintues:',
        validate: context.amplify.inputValidation({
            operator: 'regex',
            value: '^[1-9][0-9]*$', // change to /d after checking
            onErrorMsg: 'Resouce should be numeric',
            required: true,
        }),
      };
    const minuteAnswer = await inquirer.prompt([minuteQuestion]);
    const plural = minuteAnswer.minutes === 1 ? '' : 's';
    let cloudwatchRule = `rate(${minuteAnswer.minutes} minute${plural})`
    return cloudwatchRule;
}

export async function hourHelper(context : any){
    const hourQuestion = {
        type: 'input',
        name: 'hours',
        message: 'Enter the rate in hours:',
        validate: context.amplify.inputValidation({
          validation: {
            operator: 'regex',
            value: '^[1-9][0-9]*$',
            onErrorMsg: 'Resouce should be numeric',
          },
          required: true,
        }),
      };
    const hourAnswer = await inquirer.prompt([hourQuestion]);
    const plural = hourAnswer.hours === 1 ? '' : 's';
    let cloudwatchRule = `rate(${hourAnswer.hours} hour${plural})`
    return cloudwatchRule
}

export async function timeHelper( exp : CronBuilder){
    const timeQuestion = {
        type: 'datetime',
        name: 'dt',
        message: 'Select the start time (use arrow keys):',
        format: ['HH', ':', 'mm', ' ', 'A'],
      };
    const timeAnswer = await inquirer.prompt([timeQuestion]);
    exp.set(
    'minute',
    (timeAnswer.dt as any)
        .getMinutes()
        .toString()
        .split(),
    );
    exp.set(
    'hour',
    (timeAnswer.dt as any)
        .getHours()
        .toString()
        .split(),
    );
    return exp;
}

export async function weekHelper( exp : CronBuilder){
    const WeekQuestion = {
        type: 'list',
        name: 'week',
        message: 'Select the day to invoke the function:',
        choices: ['MON', 'TUE', 'WED', 'THU', 'FRI', 'SAT', 'SUN'],
    };
    const weekAnswer = await inquirer.prompt([WeekQuestion]);
    exp.set('dayOfTheWeek', Days[(weekAnswer as any).week].split());
    return exp;
}

export async function monthHelper(exp){
    const dateQuestion = {
        type: 'datetime',
        name: 'dt',
        message: 'Select on which day of the month to invoke the function (dd)(use arrow keys):',
        format: ['DD'],
      };
    const dateAnswer = await inquirer.prompt([dateQuestion]);
    exp.set(
        'dayOfTheMonth',
        (dateAnswer.dt as any)
          .getDate()
          .toString()
          .split(),
      );
    return exp;
}

export async function yearHelper(exp){
    const dateQuestion = {
        type: 'datetime',
        name: 'dt',
        message: 'Select the month and date to invoke the function (mm / dd) (use arrow keys):',
        format: ['MM', '/', 'DD'],
      };
      const dateAnswer = await inquirer.prompt([dateQuestion]);
      exp.set(
        'dayOfTheMonth',
        (dateAnswer.dt as any)
          .getDate()
          .toString()
          .split(),
      );
      exp.set(
        'month',
        (dateAnswer.dt as any)
          .getMonth()
          .toString()
          .split(),
      );
    return exp;
}