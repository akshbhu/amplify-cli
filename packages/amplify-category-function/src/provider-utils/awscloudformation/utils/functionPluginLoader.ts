import inquirer from 'inquirer'
import { FunctionParameters } from "amplify-function-plugin-interface";
import _ from 'lodash';

/*
 * This file contains the logic for loading, selecting and executing function plugins (currently runtime and template plugins)
 */

 /**
  * Loads function template plugins
  * @param context Amplify Context object
  * @param params Current function parameters
  * @returns FunctionParameters populated with the values required to build the supplied template
  */
export async function templateWalkthrough(context: any, params: FunctionParameters): Promise<FunctionParameters> {
  const pluginType = 'function-template-provider';
  const selectionItem = 'template';
  return await loadParametersFromPluginType(context, params, pluginType, selectionItem);
}

/**
 * Loads function runtime plugins
 * @param context Amplify Context object
 * @param params Current function parameters
 * @returns FunctionParameters populated with the value required to build the supplied runtime
 */
export async function runtimeWalkthrough(context: any, params: FunctionParameters): Promise<FunctionParameters> {
  const pluginType = 'function-runtime-provider';
  const selectionItem = 'runtime';
  return await loadParametersFromPluginType(context, params, pluginType, selectionItem);
}

/**
 * Handles dynamically requiring function plugins and executing the options exposed in the plugin
 * @param context Amplify Context object
 * @param params Current function parameters
 * @param pluginType The type of function plugin to load
 * @param itemName The name of the item that should be shown to the user
 */
async function loadParametersFromPluginType(context: any, params: FunctionParameters, pluginType: string, itemName: string): Promise<FunctionParameters> {
  // get providers from context
  const templateProviders = context.pluginPlatform.plugins[pluginType];
  if (!templateProviders) {
    throw 'No template plugins found. You can either create your function from scratch or download and install template plugins then rerun this command.'
  }

  const paramsCopy = _.assign({}, params) // copy params to pass to plugins so they cannot be modified

    // load the options from each provider
  const options = templateProviders
    .map(providerMeta => providerMeta.packageLocation)
    .map(packageLoc => {
        try {
          return require(packageLoc); // dynamic require each template provider
        } catch(err) {
          context.print.warning('Some plugins could not be loaded. Run "amplify plugin scan" to fix');
          return false;
        }
      }
    )
    .filter(provider => !!provider) // get rid of any cases above where we failed to load the plugin
    .flatMap(provider => provider.getOptions(context, paramsCopy)) // copy params so that plugin cannot modify them
    .map(op => {return {name: op.name, value: op.create}}); // the create function is set as the value of the selection

  // sanity checks
  if (options.length === 0) {
    context.print.warning(`No ${itemName} found for the selected function configuration`)
    context.print.warning(`You can download and install ${itemName} plugins then rerun this command`)
    return {};
  } else if (options.length === 1) {
    context.print.info(`${options[0].name} found for selected function configuration.`)
    return await options[0].value();
  }

  // ask which template to use
  const selection = await inquirer.prompt([{
    type: 'list',
    name: 'create',
    message: `Select the function ${itemName}:`,
    choices: options
  }]);

  // execute the template creation function associated with the selection
  return await selection.create();
}