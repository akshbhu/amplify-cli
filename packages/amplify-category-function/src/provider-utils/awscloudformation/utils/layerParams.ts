import { FunctionRuntime, ProviderContext } from 'amplify-function-plugin-interface';
import _ from 'lodash';
import path from 'path';

export type LayerParameters = {
  layerName: string;
  layerPermissions: Permissions[];
  runtimes: FunctionRuntime[];
  providerContext: ProviderContext;
  authorizedAccountIds?: string;
  authorizedOrgId?: string;
  layerVersionsMap?: Object;
};

export enum Permissions {
  private = 'private',
  public = 'public',
  awsAccounts = 'awsAccounts',
  awsOrg = 'awsOrg',
}

export type LayerMetadataFactory = (obj : Object) => LayerMetadata;

export interface LayerMetadata {
  runtimes: FunctionRuntime[]
  getVersion: (version: number) => LayerVersionMetadata
  listVersions:() => number[]
}

export interface LayerVersionMetadata {
  permissions: LayerPermission[]
}

export type LayerPermission = PrivateLayer | PublicLayer | AccountsLayer | OrgsLayer

export interface PrivateLayer {
  type: Permissions.private
}

export interface PublicLayer {
  type: Permissions.public
}

export interface AccountsLayer {
  type: Permissions.awsAccounts
  accounts: string[]
}

export interface OrgsLayer {
  type: Permissions.awsOrg
  orgs: string[]
}

 class LayerState implements LayerMetadata , LayerVersionMetadata{
  versionsMap : Map<number , LayerPermission[]> = new Map();
  runtimes: FunctionRuntime[];
  permissions: LayerPermission[];
  constructor(obj){
    this.runtimes = obj.runtimes;
    this.permissions =[];
    Object.entries(obj.layerVersionsMap).forEach(([key, value]) => {
      this.permissions = obj.layerVersionsMap[key];
      this.versionsMap.set(Number(key),this.permissions);
    });

  }

   getVersion(version : number) : LayerVersionMetadata{
    const obj : LayerVersionMetadata = {
      permissions : this.versionsMap.get(version),
    };
    return obj;
  }

   listVersions() : number[] {
    return Array.from(this.versionsMap.keys());
  }

}

export const layerMetadataFactory : LayerMetadataFactory = (obj : Object) : LayerMetadata => {
  const layerState : LayerState = new LayerState(obj);
  return layerState;
};

