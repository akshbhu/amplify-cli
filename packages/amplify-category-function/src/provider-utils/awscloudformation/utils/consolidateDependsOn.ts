import { $TSObject } from 'amplify-cli-core';
import { FunctionDependency } from 'amplify-function-plugin-interface';
import _ from 'lodash';
import { category } from '../../../constants';
import { lambdaLayerSetting, resourceAccessSetting, ServiceName } from './constants';

export const consolidateDependsOnForLambda = (
  projectMeta: $TSObject,
  currentDependsOn: FunctionDependency[],
  lambdaToUpdate: string,
  selectedSettings: string[],
): FunctionDependency[] => {
  let updatedDependsOn: FunctionDependency[];
  const prevFunctionParametersDependsOn: FunctionDependency[] = _.get(projectMeta, ['function', lambdaToUpdate, 'dependsOn'], []);
  if (selectedSettings.includes(resourceAccessSetting)) {
    // insert old lambdaLayer dependsOn if present
    const prevLayersDependsOn: FunctionDependency[] = prevFunctionParametersDependsOn.filter(
      resource => _.get(projectMeta, [category, resource.resourceName, 'service'], []) === ServiceName.LambdaLayer,
    );
    updatedDependsOn = currentDependsOn.concat(prevLayersDependsOn);
  }
  if (selectedSettings.includes(lambdaLayerSetting)) {
    //  insert resource dependsOn
    const prevLayersDependsOn: FunctionDependency[] = prevFunctionParametersDependsOn.filter(
      resource => _.get(projectMeta, [category, resource.resourceName, 'service'], []) !== ServiceName.LambdaLayer,
    );
    updatedDependsOn = currentDependsOn.concat(prevLayersDependsOn);
  }
  return updatedDependsOn;
};
