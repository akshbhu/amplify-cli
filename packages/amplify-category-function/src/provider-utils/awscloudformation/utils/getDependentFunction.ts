import { $TSContext, $TSObject } from 'amplify-cli-core';
import { loadFunctionParameters } from './loadFunctionParameters';
import path from 'path';
import { ServiceName } from './constants';
import { category } from '../../../constants';

export async function getDependentFunctionsCfn(
  context: $TSContext,
  allResources: $TSObject[],
  backendDir: string,
  modelsDeleted: string[],
) {
  //get the List of functions dependent on deleted model
  let functionNamesToBeUpdated = [];
  const lambdaFuncResources = allResources.filter(
    resource =>
      resource.service === ServiceName.LambdaFunction &&
      resource.mobileHubMigrated !== true &&
      resource.dependsOn.find(val => val.category === 'api'),
  );

  // initialize function parameters for update
  for (let lambda of lambdaFuncResources) {
    const resourceDirPath = path.join(backendDir, category, lambda.resourceName);
    const currentParameters = loadFunctionParameters(context, resourceDirPath);
    const selectedCategories = currentParameters.permissions;
    let deletedModelFound: boolean = true;

    for (const selectedCategory of Object.keys(selectedCategories)) {
      const selectedResources = selectedCategories[selectedCategory];
      deletedModelFound = Object.keys(selectedResources).some(r => modelsDeleted.indexOf(r) >= 0);
      if (deletedModelFound) {
        functionNamesToBeUpdated.push(lambda.resourceName);
      }
    }
  }
  return functionNamesToBeUpdated;
}
