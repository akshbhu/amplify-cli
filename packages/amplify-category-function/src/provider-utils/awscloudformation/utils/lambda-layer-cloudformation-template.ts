import { Fn, DeletionPolicy } from 'cloudform';
import _ from 'lodash';
import Lambda from 'cloudform-types/types/lambda';
import { Permissions} from './layerParams';


export default function generateLayerCfnObj(parameters) {

  const cfnObj = {
    AWSTemplateFormatVersion: '2010-09-09',
    Description: 'Lambda Layer resource stack creation using Amplify CLI',
    Parameters: {
      layerVersion:{
        Type: 'String',
        Default: '1'
      },
      env: {
        Type: "String"
      }
    },
    Resources: {
    },
    Outputs: {
      Arn: {
        Value: {
          Ref: 'LambdaLayer',
        },
      },
      Region: {
        Value: {
          Ref: 'AWS::Region',
        },
      },
    },
    Conditions: {
      HasEnvironmentParameter: {
        "Fn::Not": [
          {
            "Fn::Equals": [
                {
                    Ref: "env"
                },
                "NONE"
            ]
          }
        ]
      }
    }
  };
  const POLICY_RETAIN = DeletionPolicy.Retain;
  const layer = new Lambda.LayerVersion({
    CompatibleRuntimes: parameters.runtimes.map(runtime => runtime.cloudTemplateValue),
    Content: {
      S3Bucket: Fn.Ref('deploymentBucketName'),
      S3Key: Fn.Ref('s3Key'),
    },
    Description: parameters.description || '',
    LayerName: joinWithEnv('-', [parameters.layerName]),
    LicenseInfo: 'MIT',
  });
  layer.deletionPolicy(POLICY_RETAIN);
  _.assign(layer, { UpdateReplacePolicy: POLICY_RETAIN });

  cfnObj.Resources["LambdaLayer"] = layer;

  // parameters.laatestLayerVersion is defined
  Object.entries(parameters.layerVersionsMap).forEach(([key , value]) => {
    parameters.layerVersionsMap[key].forEach(permissions => {
      if (permissions.type === Permissions.awsAccounts) {
         assignLayerPermissionsAccounts(cfnObj,parameters,permissions.type,key);
       }
       else if (permissions.type === Permissions.awsOrg){
         assignLayerPermissionsOrgs(cfnObj,parameters,permissions.type,key);
       }
       else{
         assignLayerPermissions(cfnObj,parameters,permissions.type,key)
       }
     });
  })
  return cfnObj;
}

function assignLayerPermissions(cfnObj,parameters,permissions,version){
    // assign permissions
    const layerVersionPermissionInput = {
      Action: 'lambda:GetLayerVersion',
      LayerVersionArn: parameters.layerVersion !== version ? createLayerversionArn(parameters,version) : Fn.Ref("LambdaLayer")
    };
  let layerVersionPermission = new Lambda.LayerVersionPermission({
    ...layerVersionPermissionInput,
    Principal: '*',
  });
  layerVersionPermission.deletionPolicy(DeletionPolicy.Retain);
  let temp : string = `LambdaLayerPermission${permissions}${version}`
  if(permissions === Permissions.public){
    layerVersionPermission.Properties.Principal = "*";
    cfnObj.Resources[temp] = layerVersionPermission;
  }
  else{
    layerVersionPermission.Properties.Principal = Fn.Ref("AWS::AccountId");
    cfnObj.Resources[temp] = layerVersionPermission;
  }
}

function assignLayerPermissionsOrgs(cfnObj,parameters,permissions,version){
  // assign permissions
  let whitelistOrgIds = parameters.layerVersionsMap[version].filter(val =>  val.type === permissions)[0].orgs;
  whitelistOrgIds.forEach(org => {
    const layerVersionPermissionInput = {
      Action: 'lambda:GetLayerVersion',
      LayerVersionArn: parameters.layerVersion !== version ? createLayerversionArn(parameters,version) : Fn.Ref("LambdaLayer"),
    };
    let layerVersionPermission = new Lambda.LayerVersionPermission({
      ...layerVersionPermissionInput,
      Principal: '*',
    });
    layerVersionPermission.deletionPolicy(DeletionPolicy.Retain);
    let temp : string = `LambdaLayerPermission${permissions}${org.replace('-','')}${version}`
    layerVersionPermission.Properties.OrganizationId = `${org}`;
    cfnObj.Resources[temp] = layerVersionPermission;
  })
}

function assignLayerPermissionsAccounts(cfnObj,parameters,permissions,version){
  // assign permissions
  let whitelistAccountIds = parameters.layerVersionsMap[version].filter(val =>  val.type === permissions)[0].accounts;
  whitelistAccountIds.forEach(account => {
    const layerVersionPermissionInput = {
      Action: 'lambda:GetLayerVersion',
      LayerVersionArn: parameters.layerVersion !== version ? createLayerversionArn(parameters,version) : Fn.Ref("LambdaLayer"),
    };
    let layerVersionPermission = new Lambda.LayerVersionPermission({
      ...layerVersionPermissionInput,
      Principal: '*',
    });
    layerVersionPermission.deletionPolicy(DeletionPolicy.Retain);
    let temp : string = `LambdaLayerPermission${permissions}${account}${version}`
    layerVersionPermission.Properties.Principal = `${account}`;
    cfnObj.Resources[temp] = layerVersionPermission;
  })
}

export function joinWithEnv(separator: string, listToJoin: any[]) {
  return Fn.If(
    "HasEnvironmentParameter",
    Fn.Join(separator, [...listToJoin, Fn.Ref("env")]),
    listToJoin[0]
  );
}

export function createLayerversionArn(parameters,version){
  //arn:aws:lambda:us-west-2:136981144547:layer:layers089e3f8b-dev:1
  return Fn.Sub('arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:layer:${layerName}:${layerVersion}', {
    layerName: joinWithEnv('-', [parameters.layerName]),
    layerVersion: version
  })
}