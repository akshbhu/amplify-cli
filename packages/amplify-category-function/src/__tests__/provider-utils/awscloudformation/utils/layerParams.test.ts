import {layerMetadataFactory} from '../../../../provider-utils/awscloudformation/utils/layerParams';

const layerObj1 = {
    runtimes: [
        {
            name: "Java",
            value: "java",
            cloudTemplateValue: "java11",
            defaultHandler: "example.LambdaRequestHandler::handleRequest",
            layerExecutablePath: "bin/java/lib/"
        }
    ],
    layerVersionsMap: 
    {
        1: 
        [
            {
              type : "Permissions.public"
            }
        ],
        2: 
        [
          {
            type : "Permissions.awsOrg",
            orgs : [" o-dsjnkdjs"," o-sdnckjsd"]

          },
          {
            type : "Permissions.awsAccounts",
            accounts : ["11111111111","222222222"]
          }
        ],
        3: 
        [
          {
            type : "Permissions.private"
          }
        ]
    }
}

const layerObj2 = {
    runtimes: [
        {
            name: "Java",
            value: "java",
            cloudTemplateValue: "java11",
            defaultHandler: "example.LambdaRequestHandler::handleRequest",
            layerExecutablePath: "bin/java/lib/"
        }
    ],
    layerVersionsMap: 
    {
    }
}

describe('layer metadata Factory Function', () => {

  it('getversion permissions happy case 1', () => {
        layerMetadataFactory(layerObj1).listVersions().forEach(version => {
        expect(layerMetadataFactory(layerObj1).getVersion(version).permissions).toMatchSnapshot();
      });
  });
  it('getversion permissions happy case 2', () => {
        layerMetadataFactory(layerObj2).listVersions().forEach(version => {
        expect(layerMetadataFactory(layerObj2).getVersion(version).permissions).toMatchSnapshot();
      });
  });

  it('Listversion happy case 1', () => {
    expect(layerMetadataFactory(layerObj1).listVersions()).toMatchSnapshot();
  });

  it('Listversion happy case 2', () => {
    expect(layerMetadataFactory(layerObj2).listVersions()).toMatchSnapshot()
  });
});
