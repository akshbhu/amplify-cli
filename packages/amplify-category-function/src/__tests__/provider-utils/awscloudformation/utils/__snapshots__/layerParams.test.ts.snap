// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layer metadata Factory Function Listversion happy case 1 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`layer metadata Factory Function Listversion happy case 2 1`] = `Array []`;

exports[`layer metadata Factory Function getversion permissions happy case 1 1`] = `
Array [
  Object {
    "type": "Permissions.public",
  },
]
`;

exports[`layer metadata Factory Function getversion permissions happy case 1 2`] = `
Array [
  Object {
    "orgs": Array [
      " o-dsjnkdjs",
      " o-sdnckjsd",
    ],
    "type": "Permissions.awsOrg",
  },
  Object {
    "accounts": Array [
      "11111111111",
      "222222222",
    ],
    "type": "Permissions.awsAccounts",
  },
]
`;

exports[`layer metadata Factory Function getversion permissions happy case 1 3`] = `
Array [
  Object {
    "type": "Permissions.private",
  },
]
`;
